### PR TITLE
feat: use angular compiler api to transform codes

### LINKED ISSUE #108: Issue with ReferenceError - Please help.
Hi,

I'm not sure if this is an issue with Jest, Angular, jest-preset-angular, tsconfig.spec.ts or just something stupid I'm doing. I'm hoping someone can help me with this as I've been going round in circles without getting closer to the issue. I don't seem to know anyone with the same issue.

I have an angular module that when I include in a test gives me errors. The components / classes that are throwing the error are not even being tested at the moment and the module works fine in the application.

Here is a brief example of what I think is causing the issue:

`@Component(...)
export class ChildA {
  constructor(
    private parent: ParentB
  ) {}
}

@Component(...) 
export class ParentB {
}`

I understand that the declaration is above the definition in the file. However, it only throws an error in the tests. All my other modules and components are fine.

Including the module with these components throws the following error in my tests:
'ReferenceError: ParentB is not defined`

I'm hoping someone has faced a similar issue or can at least point me in the right direction.

### LINKED ISSUE #288: Angular 8: Can't resolve all parameters for Component: (?).
I recently upgraded my application to Angular 8 and couldn't get the tests to run successfully with _Jest-Preset-Angular_ resulting in the error below telling me it cannot resolve the constructor dependencies although it works fine using _Karma_ and was working with Angular 7.

```
Can't resolve all parameters for AppComponent: (?).
```

I've created a newly generated Angular 8 application to see if it has something to do with my migration and I ended up having the same issue.

Here is the reproduction repository: https://github.com/jfcere/jest-preset-angular-issue

### LINKED ISSUE #322: Can't resolve All parameters
Hi,
I recently changed to jest. Still trying to learn and optimize the testing process for our app. But switching to jest broke some tests. These tests keep failing and I can't figure why they fail. Maybe I'm doing something wrong or this is a bug.

I have already looked at similar issues but I think they are different from mine.  #154 #288 
I also used the https://github.com/briebug/jest-schematic shematic to do all the config setup. Which automatically added the emitDecoratorMetadata so I am excluding that as a possible issue.

**The component I'm trying to test**
This component just has one dependency _GoogleChartComponent_ which is a part of the [_Ng2GoogleChartsModule_](https://github.com/gmazzamuto/ng2-google-charts). 

```typescript
import { Component, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
import { IColumnRole } from '@influo/modules/chart/chart.helpers';
import { GoogleChartInterface } from 'ng2-google-charts/google-charts-interfaces';
import { BehaviorSubject } from 'rxjs';

@Component({
  selector: 'influo-geo-chart',
  template: `
      <google-chart
              *ngIf="googleChartConfig$ | async as googleChartConfig; else loader"
              [data]="googleChartConfig">
      </google-chart>
      <ng-template #loader>
          <mat-spinner color="accent" diameter="80" strokeWidth="8"></mat-spinner>
      </ng-template>
  `,
  styleUrls: [ './geo-chart.component.scss' ]
})
export class GeoChartComponent implements OnInit, OnChanges {
  @Input() columns: Array<IColumnRole | string>;
  @Input() config: GoogleChartInterface;
  @Input() data: Array<Array<string | number>>;

  private defaultConfig: GoogleChartInterface = {
    chartType: 'GeoChart',
    dataTable: [],
    options: {
      legend: false,
      region: 155,
      enableRegionInteractivity: true,
      displayMode: 'region',
      colors: [ '#e6e6e6', '#1672AD' ],
      datalessRegionColor: '#e6e6e6',
    },
  };

  constructor() {
  }

  _googleChartConfig = new BehaviorSubject<GoogleChartInterface | null>(null);

  set googleChartConfig(config: GoogleChartInterface) {
    const value = this._googleChartConfig.getValue() || {};

    this._googleChartConfig.next(Object.assign({}, value, config));
  }

  get googleChartConfig$() {
    return this._googleChartConfig.asObservable();
  }

  ngOnInit() {
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (this.columns && this.data) {
      this.googleChartConfig = Object.assign({}, this.defaultConfig, this.config, {
        dataTable: [
          this.columns,
          ...this.data,
        ],
      });
    }
  }

}

```

A very basic test. I am encapsulating _GoogleChartComponent_.
```typescript

import { CommonModule } from '@angular/common';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { UiKitModule } from '@influo/modules/ui-kit/ui-kit.module';
import { Ng2GoogleChartsModule } from 'ng2-google-charts';

import { GeoChartComponent } from './geo-chart.component';

describe('GeoChartComponent', () => {
  let component: GeoChartComponent;
  let fixture: ComponentFixture<GeoChartComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
        imports: [
          CommonModule,
          Ng2GoogleChartsModule,
          UiKitModule
        ],
        providers: [],
        declarations: [ GeoChartComponent ]
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(GeoChartComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
```
**Error**
When I look at the constructor of the GoogleChartComponent I find _ElementRef, GoogleChartsLoaderService_. _GoogleChartsLoaderService_ is provided by the module itself. So it's not possible that it's not provided. 

When I use this module outside of a testing environment it works perfectly. So my hunch is that it has to do something with the use of jsdom but this seems wrong to assume that.
Which confuses me more why this error is even occurring. 

```
Error: Can't resolve all parameters for GoogleChartComponent: (?, ?). 


    at syntaxError (/Users/other/Documents/projects/influo/src/main/packages/compiler/src/util.ts:100:17)
    at CompileMetadataResolver._getDependenciesMetadata (/Users/other/Documents/projects/influo/src/main/packages/compiler/src/metadata_resolver.ts:957:27)
...
```
Thanks in advance!

### LINKED ISSUE #353: entryComponents with Angular 9 (Ivy)
With Angular 9 (Ivy), we no longer have to provide `entryComponents`.
This seems to fail if we're using Jest as our test runner.
The same test does pass with Jasmine/Karma.

If we run it with Jest, we're getting the following error

```
No component factory found for DynamicButtonComponent. Did you add it to @NgModule.entryComponents?

      at noComponentFactoryError (../packages/core/src/linker/component_factory_resolver.ts:17:17)
      at CodegenComponentFactoryResolver.Object.<anonymous>.CodegenComponentFactoryResolver.resolveComponentFactory (../packages/core/src/linker/component_factory_resolver.ts:72:13)
      at NgComponentOutlet.Object.<anonymous>.NgComponentOutlet.ngOnChanges (../packages/common/src/directives/ng_component_outlet.ts:105:36)
      at checkAndUpdateDirectiveInline (../packages/core/src/view/provider.ts:207:15)
      at checkAndUpdateNodeInline (../packages/core/src/view/view.ts:429:14)
      at checkAndUpdateNode (../packages/core/src/view/view.ts:389:12)
      at debugCheckAndUpdateNode (../packages/core/src/view/services.ts:430:44)
      at debugCheckDirectivesFn (../packages/core/src/view/services.ts:391:7)
      at Object.eval [as updateDirectives] (ng:/DynamicHostModule/DynamicHostComponent.ngfactory.js:14:9)
      at Object.debugUpdateDirectives [as updateDirectives] (../packages/core/src/view/services.ts:385:19)
      at checkAndUpdateView (../packages/core/src/view/view.ts:359:12)
      at callViewAction (../packages/core/src/view/view.ts:615:11)
      at execComponentViewsAction (../packages/core/src/view/view.ts:559:7)
      at checkAndUpdateView (../packages/core/src/view/view.ts:370:3)
      at callWithDebugContext (../packages/core/src/view/services.ts:629:23)
      at Object.debugCheckAndUpdateView [as checkAndUpdateView] (../packages/core/src/view/services.ts:346:10)
      at ViewRef_.Object.<anonymous>.ViewRef_.detectChanges (../packages/core/src/view/refs.ts:260:16)
      at ComponentFixture.Object.<anonymous>.ComponentFixture._tick (../../packages/core/testing/src/component_fixture.ts:107:28)
      at ../../packages/core/testing/src/component_fixture.ts:120:36
      at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:396:30)
      at ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:117:43)
      at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:395:36)
      at Object.onInvoke (../packages/core/src/zone/ng_zone.ts:302:25)
      at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:395:36)
      at Zone.run (node_modules/zone.js/dist/zone.js:153:47)
      at NgZone.Object.<anonymous>.NgZone.run (../packages/core/src/zone/ng_zone.ts:178:50)
      at ComponentFixture.Object.<anonymous>.ComponentFixture.detectChanges (../../packages/core/testing/src/component_fixture.ts:120:19)
      at src/app/dynamic-host.component.spec.ts:15:13
      at node_modules/tslib/tslib.js:113:75
      at new ZoneAwarePromise (node_modules/zone.js/dist/zone.js:915:33)
      at Object.__awaiter (node_modules/tslib/tslib.js:109:16)
      at src/app/dynamic-host.component.spec.ts:13:49
      at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:396:30)
      at ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:117:43)
      at ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:395:36)
      at Zone.run (node_modules/zone.js/dist/zone.js:153:47)
```

See https://github.com/timdeschryver/jest-dynamic-ng-repro for a reproduction.

### LINKED ISSUE #622: Missing dependency, and other failures with 9.0.0-next.2
## 🐛 Bug Report

While testing Ivy support using 9.0.0-next.2, I ran into some significant issues. They are as follows:

Background: We have a medium-sized project with a few hundred tests using jest-preset-angular for some time. It does not have many external dependencies by most Angular project standards.

1. After first upgrading to `jest-preset-angular@9.0.0-next.2`, I received this error for every test:

```
 FAIL  libs/ui-components/dialog/src/dialog-close.directive.spec.ts
  ● Test suite failed to run
    
	Cannot find module '@ngtools/webpack/src/diagnostics'
    Require stack:
    - C:\src\ease\WebUi\node_modules\jest-preset-angular\build\compiler\ng-jest-compiler.js
    - C:\src\ease\WebUi\node_modules\jest-preset-angular\build\index.js
    - C:\src\ease\WebUi\node_modules\@jest\transform\build\ScriptTransformer.js
    - C:\src\ease\WebUi\node_modules\@jest\transform\build\index.js
    - C:\src\ease\WebUi\node_modules\jest-runtime\build\index.js
    - C:\src\ease\WebUi\node_modules\jest-runner\build\testWorker.js
    - C:\src\ease\WebUi\node_modules\jest-worker\build\workers\processChild.js

      at Object.<anonymous> (node_modules/jest-preset-angular/build/compiler/ng-jest-compiler.js:6:21)
```

To fix this, I tried:

```
> yarn add -D @ngtools/webpack

...
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 1 new dependency.
info Direct dependencies
└─ @ngtools/webpack@10.2.0
info All dependencies
└─ @ngtools/webpack@10.2.0	 
```

This may be the cause of the following problems.

2. Now, while running tests, I experienced a "hang" for 8+ minutes before any tests were run. It may have been longer - I tried attaching a debugger but wasn't able to. After this, the starting "hang" seems shorter - a minute or two.

Eventually some tests were run, but I'm seeing a large number of odd compile + runtime errors. A single digit number of tests passed successfully.

The main compile error I'm seeing is:
```
    libs/ui-components/form/src/text-field/text-field.component.ts:50:43 - error TS2340: Only public and protected methods of the base class are accessible via the 'super' keyword.

    50     return this._input?.disabled ?? super.disabled;
```

Runtime errors include a lot of these (intermixed):

1:
```
Failed: "Zone is needed for the waitForAsync() test helper but could not be found. Please make sure that your environment includes zone.js/dist/zone.js"

      17 |   let intersectionObserverService: IntersectionObserverServiceMock;
      18 | 
    > 19 |   beforeEach(waitForAsync(() => {
         |   ^
      20 |     configureTestBed(TestDefaultComponent);
      21 | 
      22 |     fixture = createTestComponent(TestDefaultComponent);

      at Env.beforeEach (node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:46:24)
```

2:
```

    TypeError: Cannot read property 'debugElement' of undefined

      181 | 
      182 |   it('should unsubscribe from intersection observer', () => {
    > 183 |     const button = fixture.debugElement.query(By.directive(VisibilityTrackingDirective));
          |                            ^
      184 |     expect(button).toBeDefined();
      185 | 
      186 |     showElement(button, 1000, 0.1, intersectionObserverService);
```

## To Reproduce

Let me know if you'd like me to create a public repro with a few tests exhibiting these problems. I'm happy to, but wanted to start with a report.

## envinfo

```
System:
    OS: Windows 10

Npm packages:
   jest@26.6.3
   jest-preset-angular@9.0.0-next.2
   typescript@4.0.5
```

Related to [#409](409)

### PR DESCRIPTION
**Pull request by ahnpnl**
## Summary

See individual commits

`NgJestCompiler` takes care of compiling files and return a string which is the compiled output concatenating with source map string. The compiler also does type checking.

This compiler is mainly inspired by [AngularCompilerPlugin](https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/angular_compiler_plugin.ts) in combination with `ts-jest` [LanguageService](https://github.com/kulshekhar/ts-jest/blob/master/src/compiler/language-service.ts).

Closes #108 
Closes #288
Closes #322 
Closes #353 
Closes #622  

### Check list

- [x] Type checking

- [x] Code compilation

- [x] Memory cache for `Program`

- [x] Allow custom compiler module, e.g `ttypescript` (via `ts-jest` option)  

- [x] Downlevel ctor transformer

- [x] hoisting (from `ts-jest`) 

- [x] `ngcc` util script

## Test plan

- Added unit tests

- Added e2e test which is taken from #353

- Added e2e test for `forwardRef`

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

With the new jest transformer, `jest-preset-angular` now switches to default to use this new transformer and no longer use `ts-jest` to transform codes.

Users who are currently doing in jest config
```
// jest.config.js
module.exports = {
    // [...]
    transform: {
      '^.+\\.(ts|js|html)$': 'ts-jest',
    },
}
```

should change to
```
// jest.config.js
module.exports = {
    // [...]
    transform: {
      '^.+\\.(ts|js|html)$': 'jest-preset-angular',
    },
}
```

`isolatedModule: true` will still use `ts-jest` to compile `ts` to `js` but you won't get full compatibility with Ivy.

## Other information
**N.A.**

### CODE REVIEW

**Comment by ahnpnl** [comment]
@wtho here we go !!! The compiler now works with our current e2e tests, as well as solves #353

File: jest-preset.js [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
       tsconfig: '<rootDir>/tsconfig.spec.json',
       stringifyContentPathRegex: '\\.html$',
-      astTransformers: {
-        before: customTransformers,
-      },
     },
   },
   transform: {
-    '^.+\\.(ts|js|html)$': 'ts-jest',
+    '^.+\\.(ts|js|html)$': 'jest-preset-angular',
```
Comment (wtho) [inline jest-preset.js:11]: Wow this is it! Haha, will have to get used to it!

**Review by wtho — COMMENTED** [review]
And you added the `downlevelConstructorParameters`!

Wow this is so cool! I wish I would have been more involved into this, but I was quite busy recently.
Once more thanks again!

**Comment by ahnpnl** [comment]
Still have some fine tunings to do but we are pretty close. The current state is already in pretty good shape. Probably next week I can push this out in 9.0.0-next.1 once this one is done.

**Comment by wtho** [comment]
What's still open? Is there still more to address?

**Comment by ahnpnl** [comment]
Should be ready now, I couldn't find example in #322 to try out, but seems like #322 will be solved with this PR too.

**Comment by ahnpnl** [comment]
I have published this PR under **9.0.0-next.1** release so it's easier to test with real projects.

File: e2e/test-app-v10-zone-v11/jest.config.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require('jest-preset-angular/ngcc-jest-processor');
```
Comment (ahnpnl) [inline e2e/test-app-v10-zone-v11/jest.config.js:1]: @wtho this is the only way

**Pull request merged**